components/core/src/clp_s/ArchiveWriter.cpp

```diff
@@ -138,19 +142,70 @@ size_t ArchiveWriter::store_tables() {
             FileWriter::OpenMode::CreateForWriting
     );
     m_table_metadata_compressor.open(m_table_metadata_file_writer, m_compression_level);
-    m_table_metadata_compressor.write_numeric_value(m_id_to_schema_writer.size());
-    for (auto& i : m_id_to_schema_writer) {
-        m_table_metadata_compressor.write_numeric_value(i.first);
-        m_table_metadata_compressor.write_numeric_value(i.second->get_num_messages());
-        m_table_metadata_compressor.write_numeric_value(m_tables_file_writer.get_pos());
 
-        m_tables_compressor.open(m_tables_file_writer, m_compression_level);
-        size_t uncompressed_size = i.second->store(m_tables_compressor);
-        m_tables_compressor.close();
-        delete i.second;
+    using schema_map_it = decltype(m_id_to_schema_writer)::iterator;
+    std::vector<schema_map_it> schemas;
+    std::vector<std::tuple<size_t, size_t>> table_metadata;
+    std::vector<std::tuple<int32_t, size_t, size_t, size_t>> schema_metadata;
 
+    schema_metadata.reserve(m_id_to_schema_writer.size());
+    schemas.reserve(m_id_to_schema_writer.size());
+    auto it = m_id_to_schema_writer.begin();
+    for (size_t i = 0; i < m_id_to_schema_writer.size(); ++i) {
+        schemas.push_back(it++);
+    }
+    auto comp = [](schema_map_it const& lhs, schema_map_it const& rhs) -> bool {
+        return lhs->second->get_total_uncompressed_size()
+               > rhs->second->get_total_uncompressed_size();
+    };
+    std::sort(schemas.begin(), schemas.end(), comp);
+
+    size_t current_table_size = 0;
```

Comment:
Do you think we should come up with a better name since in line 203, we use `table_offset`?

Reply:
I'll change it to current_stream_size, and stream_offset.